Remove functor factories in KernelFunction

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35488 Remove functor factories in KernelFunction**
* #36394 Remove experimental c10 ops

-
The original problem why those existed was a SIOF (see the multi-line comment that is deleted in this PR).
However, I think this SIOF situation only happened for caffe2 kernels exposed to PyTorch and those now use a different mechanism that shouldn't cause the SIOF anymore (they now create the caffe2 kernel instance on each call instead of storing it in the functor). If this PR passes CI, I'm assuming that the SIOF doesn't exist anymore and we can simplify this code.

Differential Revision: [D20676093](https://our.internmc.facebook.com/intern/diff/D20676093/)

smessmer committed:
Remove functor factories in KernelFunction

-
The original problem why those existed was a SIOF (see the multi-line comment that is deleted in this PR).
However, I think this SIOF situation only happened for caffe2 kernels exposed to PyTorch and those now use a different mechanism that shouldn't cause the SIOF anymore (they now create the caffe2 kernel instance on each call instead of storing it in the functor). If this PR passes CI, I'm assuming that the SIOF doesn't exist anymore and we can simplify this code.

Differential Revision: [D20676093](https://our.internmc.facebook.com/intern/diff/D20676093/)

[ghstack-poisoned]

bhosmer commented:
Are the test failures unrelated to the SIOF?

smessmer committed:
Update on "Remove functor factories in KernelFunction"


-
The original problem why those existed was a SIOF (see the multi-line comment that is deleted in this PR).
However, I think this SIOF situation only happened for caffe2 kernels exposed to PyTorch and those now use a different mechanism that shouldn't cause the SIOF anymore (they now create the caffe2 kernel instance on each call instead of storing it in the functor). If this PR passes CI, I'm assuming that the SIOF doesn't exist anymore and we can simplify this code.

Differential Revision: [D20676093](https://our.internmc.facebook.com/intern/diff/D20676093/)

[ghstack-poisoned]

bhosmer reviewed (approved):
Nice simplification! Accepting mod SIOF-induced test failures 😬

smessmer committed:
Update on "Remove functor factories in KernelFunction"


-
The original problem why those existed was a SIOF (see the multi-line comment that is deleted in this PR).
However, I think this SIOF situation only happened for caffe2 kernels exposed to PyTorch and those now use a different mechanism that shouldn't cause the SIOF anymore (they now create the caffe2 kernel instance on each call instead of storing it in the functor). If this PR passes CI, I'm assuming that the SIOF doesn't exist anymore and we can simplify this code.

Differential Revision: [D20676093](https://our.internmc.facebook.com/intern/diff/D20676093/)

[ghstack-poisoned]

ezyang reviewed (approved):
It'd be cool if we could get rid of the `unique_ptr` too but this seems like a good start.

smessmer committed:
Update on "Remove functor factories in KernelFunction"


-
The original problem why those existed was a SIOF (see the multi-line comment that is deleted in this PR).
However, I think this SIOF situation only happened for caffe2 kernels exposed to PyTorch and those now use a different mechanism that shouldn't cause the SIOF anymore (they now create the caffe2 kernel instance on each call instead of storing it in the functor). If this PR passes CI, I'm assuming that the SIOF doesn't exist anymore and we can simplify this code.

Differential Revision: [D20676093](https://our.internmc.facebook.com/intern/diff/D20676093/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove functor factories in KernelFunction"


-
The original problem why those existed was a SIOF (see the multi-line comment that is deleted in this PR).
However, I think this SIOF situation only happened for caffe2 kernels exposed to PyTorch and those now use a different mechanism that shouldn't cause the SIOF anymore (they now create the caffe2 kernel instance on each call instead of storing it in the functor). If this PR passes CI, I'm assuming that the SIOF doesn't exist anymore and we can simplify this code.

Differential Revision: [D20676093](https://our.internmc.facebook.com/intern/diff/D20676093/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove functor factories in KernelFunction"


-
The original problem why those existed was a SIOF (see the multi-line comment that is deleted in this PR).
However, I think this SIOF situation only happened for caffe2 kernels exposed to PyTorch and those now use a different mechanism that shouldn't cause the SIOF anymore (they now create the caffe2 kernel instance on each call instead of storing it in the functor). If this PR passes CI, I'm assuming that the SIOF doesn't exist anymore and we can simplify this code.

Differential Revision: [D20676093](https://our.internmc.facebook.com/intern/diff/D20676093/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove functor factories in KernelFunction"


-
The original problem why those existed was a SIOF (see the multi-line comment that is deleted in this PR).
However, I think this SIOF situation only happened for caffe2 kernels exposed to PyTorch and those now use a different mechanism that shouldn't cause the SIOF anymore (they now create the caffe2 kernel instance on each call instead of storing it in the functor). If this PR passes CI, I'm assuming that the SIOF doesn't exist anymore and we can simplify this code.

Differential Revision: [D20676093](https://our.internmc.facebook.com/intern/diff/D20676093/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove functor factories in KernelFunction"


-
The original problem why those existed was a SIOF (see the multi-line comment that is deleted in this PR).
However, I think this SIOF situation only happened for caffe2 kernels exposed to PyTorch and those now use a different mechanism that shouldn't cause the SIOF anymore (they now create the caffe2 kernel instance on each call instead of storing it in the functor). If this PR passes CI, I'm assuming that the SIOF doesn't exist anymore and we can simplify this code.

Differential Revision: [D20676093](https://our.internmc.facebook.com/intern/diff/D20676093/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove functor factories in KernelFunction"


-
The original problem why those existed was a SIOF (see the multi-line comment that is deleted in this PR).
However, I think this SIOF situation only happened for caffe2 kernels exposed to PyTorch and those now use a different mechanism that shouldn't cause the SIOF anymore (they now create the caffe2 kernel instance on each call instead of storing it in the functor). If this PR passes CI, I'm assuming that the SIOF doesn't exist anymore and we can simplify this code.

Differential Revision: [D20676093](https://our.internmc.facebook.com/intern/diff/D20676093/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove functor factories in KernelFunction"


-
The original problem why those existed was a SIOF (see the multi-line comment that is deleted in this PR).
However, I think this SIOF situation only happened for caffe2 kernels exposed to PyTorch and those now use a different mechanism that shouldn't cause the SIOF anymore (they now create the caffe2 kernel instance on each call instead of storing it in the functor). If this PR passes CI, I'm assuming that the SIOF doesn't exist anymore and we can simplify this code.

Differential Revision: [D20676093](https://our.internmc.facebook.com/intern/diff/D20676093/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove functor factories in KernelFunction"


-
The original problem why those existed was a SIOF (see the multi-line comment that is deleted in this PR).
However, I think this SIOF situation only happened for caffe2 kernels exposed to PyTorch and those now use a different mechanism that shouldn't cause the SIOF anymore (they now create the caffe2 kernel instance on each call instead of storing it in the functor). If this PR passes CI, I'm assuming that the SIOF doesn't exist anymore and we can simplify this code.

Differential Revision: [D20676093](https://our.internmc.facebook.com/intern/diff/D20676093/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove functor factories in KernelFunction"


-
The original problem why those existed was a SIOF (see the multi-line comment that is deleted in this PR).
However, I think this SIOF situation only happened for caffe2 kernels exposed to PyTorch and those now use a different mechanism that shouldn't cause the SIOF anymore (they now create the caffe2 kernel instance on each call instead of storing it in the functor). If this PR passes CI, I'm assuming that the SIOF doesn't exist anymore and we can simplify this code.

Differential Revision: [D20676093](https://our.internmc.facebook.com/intern/diff/D20676093/)

[ghstack-poisoned]

smessmer committed:
Update on "Remove functor factories in KernelFunction"


-
The original problem why those existed was a SIOF (see the multi-line comment that is deleted in this PR).
However, I think this SIOF situation only happened for caffe2 kernels exposed to PyTorch and those now use a different mechanism that shouldn't cause the SIOF anymore (they now create the caffe2 kernel instance on each call instead of storing it in the functor). If this PR passes CI, I'm assuming that the SIOF doesn't exist anymore and we can simplify this code.

Differential Revision: [D20676093](https://our.internmc.facebook.com/intern/diff/D20676093/)

[ghstack-poisoned]
